gelf/reader_test.go

```diff
@@ -0,0 +1,233 @@
+package gelf
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+)
+
+func TestReader(t *testing.T) {
+	notChunked := createMessage("a", 10)
+	notChunkedBytes := prepareUDPMessages(t, notChunked)
+	//message that is sent in 2 chunks
+	chunked1 := createMessage("b", ChunkSize+1)
+	chunkedBytes1 := prepareUDPMessages(t, chunked1)
+	//message that is sent in 3 chunks
+	chunked2 := createMessage("c", ChunkSize*2)
+	chunkedBytes2 := prepareUDPMessages(t, chunked2)
+	//message that is sent in 129 chunks
+	messageBytes3 := prepareUDPMessages(t, createMessage("d", chunkedDataLen*128))
+	//message that is sent in 128 chunks
+	chunked4 := createMessage("e", chunkedDataLen*127)
+	chunkedBytes4 := prepareUDPMessages(t, chunked4)
+
+	tests := map[string]struct {
+		udpMessagesToSend [][]byte
+		expectedMessages  []Message
+		expectedErrors    []error
+	}{
+		"expected messages to be received even if chunked messages are received in parallel": {
+			udpMessagesToSend: [][]byte{chunkedBytes1[0], chunkedBytes2[0], chunkedBytes2[1], chunkedBytes1[1], chunkedBytes2[2]},
+			expectedMessages:  []Message{chunked1, chunked2},
+		},
+		"expected message to be received even if chunks are sent in any order": {
+			udpMessagesToSend: [][]byte{chunkedBytes2[1], chunkedBytes2[0], chunkedBytes2[2]},
+			expectedMessages:  []Message{chunked2},
+		},
+		"expected messages to be received even if not chunked message is sent between chunks": {
+			udpMessagesToSend: [][]byte{chunkedBytes1[0], notChunkedBytes[0], chunkedBytes1[1]},
+			expectedMessages:  []Message{chunked1, notChunked},
+		},
+		"expected not chunked messages to be received": {
+			udpMessagesToSend: [][]byte{notChunkedBytes[0], notChunkedBytes[0]},
+			expectedMessages:  []Message{notChunked, notChunked},
+		},
+		"expected error if chunked message length is not greater than 12 bytes": {
+			udpMessagesToSend: [][]byte{magicChunked},
+			expectedErrors:    []error{fmt.Errorf("chunked message size must be greather than 12")},
+		},
+		"expected error if message is split to more than 128 chunks": {
+			udpMessagesToSend: messageBytes3,
+			expectedErrors:    []error{fmt.Errorf("message must not be split into more than %v chunks", maxChunksCount)},
+		},
+		"expected messages to be received even if it's split into 128 chunks": {
+			udpMessagesToSend: chunkedBytes4,
+			expectedMessages:  []Message{chunked4},
+		},
+	}
+
+	for name, data := range tests {
+		t.Run(name, func(t *testing.T) {
+			reader, err := NewReader(":0")
+			require.NoError(t, err)
+			done := make(chan bool)
+			defer func() {
+				_ = reader.Close()
```

Comment:
Nit: If you want to check the error here you could do this:
```suggestion
				require.NoError(t, reader.Close())
```

Reply:
done